CONTRIBUTING.md

```diff
@@ -0,0 +1,121 @@
+  # only if you have upx installed
+  upx -7 -k target/clickhouse-jdbc-bin
+  ```
+
+- run native binary
+
+  ```bash
+  # print usage
+  ./target/clickhouse-jdbc-bin
+  Usage: clickhouse-jdbc-bin [PROPERTIES] <URL> [QUERY] [FILE]
+  ...
+
+  # test database connection using JDBC driver
+  ./target/clickhouse-jdbc-bin -Dverbose=true 'jdbc:ch:http://localhost'
+  Arguments:
+    -   url=jdbc:ch:http://localhost
+    - query=select 500000000
+    -  file=jdbc.out
+
+  Options: action=read, batch=1000, samples=500000000, serde=true, type=, verbose=true
+  Processed 1 rows in 85.56 ms (11.69 rows/s)
+
+  # test query performance using Java client
+  ./target/clickhouse-jdbc-bin -Dverbose=true -Dtype=uint64 'http://localhost'
+  ...
+  Processed 500,000,000 rows in 52,491.38 ms (9,525,373.89 rows/s)
+
+  # test same query again using JVM for comparison - don't have GraalVM EE so JIT wins in my case
+  java -Dverbose=true -Dtype=uint64 -jar target/clickhouse-jdbc-*-http.jar 'http://localhost'
+  ...
+  Processed 500,000,000 rows in 25,267.89 ms (19,787,963.94 rows/s)
+  ```
+
+## Testing
+
+By default, [docker](https://docs.docker.com/engine/install/) is required to run integration test. Docker image(defaults to `clickhouse/clickhouse-server`) will be pulled from Internet, and containers will be created automatically by [testcontainers](https://www.testcontainers.org/) before testing. To test against specific version of ClickHouse, you can pass parameter like `-DclickhouseVersion=22.8` to Maven.
+
+In the case you don't want to use docker and/or prefer to test against an existing server, please follow instructions below:
+
+- make sure the server can be accessed using default account(user `default` and no password), which has both DDL and DML privileges
+- add below two configuration files to the existing server and expose all defaults ports for external access
+  - [ports.xml](../../blob/main/clickhouse-client/src/test/resources/containers/clickhouse-server/config.d/ports.xml) - enable all ports
+  - and [users.xml](../../blob/main/clickhouse-client/src/test/resources/containers/clickhouse-server/users.d/users.xml) - accounts used for integration test
+    Note: you may need to change root element from `clickhouse` to `yandex` when testing old version of ClickHouse.
+- make sure ClickHouse binary(usually `/usr/bin/clickhouse`) is available in PATH, as it's required to test `clickhouse-cli-client`
+- put `test.properties` under either `~/.clickhouse` or `src/test/resources` of your project, with content like below:
+  ```properties
+  clickhouseServer=x.x.x.x
+  # below properties are only useful for test containers
+  #clickhouseVersion=latest
+  #clickhouseTimezone=UTC
+  #clickhouseImage=clickhouse/clickhouse-server
+  #additionalPackages=
+  ```
+
+### Tooling
+We use [TestNG](http://org.testng/doc) as testing framework and for running ClickHouse Local instance [testcontainers](https://www.testcontainers.org/modules/databases/clickhouse/).
```

Comment:
Should we format the doc so that there will be a blank line between header and paragraph?